### ISSUE TITLE
Dependencies for shared lib compilations

### ISSUE THREAD

**Issue by dl5rcw** [mxe/mxe#1180]
Dear community, Dear developers,

I started to use shared libraries when compiling projects. 
Running Jenkins + mxe on a Debian 64bit server, I use nsis to create <project>_<svn-revision>_<mxe-compiler>.exe Install files right away, copying them as nightly builds on my apache. 

Since I use shared libraries now, I need to figure out which packages/libraries/(*.dll files) to add to my nsis installer. I am running on Debian 64 bit, so I need to figure out which libraries are actually needed by the created *exe file.

I messed around with 
   wine, 
   winetricks, 
   wine + dependencywalker, etc. 
But still found no solution yet. Is there any feature included in mxe that I am missing out on? Has anyone ideas how we could enhance this, so shared libraries on mxe + nsis get the final kick out of it for our community?

Best regards,
   Lars

**Comment by starius** [comment]
As a part of [MXE binary distribution](http://pkg.mxe.cc), lists of all files of each package are provided. I suppose you already know which MXE packages you use. Download and unpack archive [list.tar.xz](http://pkg.mxe.cc/repos/list.tar.xz) and find all dll files in packages you use.

Example. Finding all dll files of package `ucl` in `i686-w64-mingw32.shared` target:
```
$ grep -i dll$ mxe-i686-w64-mingw32.shared-ucl.list
usr/i686-w64-mingw32.shared/bin/libucl-1.dll
```

**Comment by tonytheodore** [comment]
> Since I use shared libraries now, I need to figure out which packages/libraries/(*.dll files) to add to my nsis installer.

For deployment, I think something like [tools/copydlldeps.py](https://github.com/mxe/mxe/blob/master/tools/copydlldeps.py) is what you're after. However, it has some issues (#589 #564) and there is an alternate approach mentioned in https://github.com/mxe/mxe/pull/530#issuecomment-62293374.

I haven't tried either, but I think an `objdump` or `nm` based approach would likely be more robust.

**Comment by dl5rcw** [comment]
Thanks for your reply.  
I am looking for a solution where I can start from the created *exe file. I have currently seven projects that compile in this environment, each using a subset of my MXE overall package. Going from the final exe would make this code segment interesting for each project in the built environment. All of your responses sound great and are indeed inspiring me for a bunch of additional options again!

As a coincidence, my built server boke yesterday. Recovering with ReaR and BareOS scheduled for Saturday, but I will need to recompile MXE, too.
Give me a week, but be assured I will give it a try and let you know.
The objdump - tried two weeks ago - seemed not to grep all dependencies of a qt5 project though. Don't know if this is an issue about recurrence?

**Comment by dl5rcw** [comment]
Dear Community,

I think tools/copydlldeps.py is just what is needed here! Unfortunately, it failed. 
For project Nr 1, it looks like an encoding issue (ascii vs utf-8) - this my current starting point.
On a second project of mine, I can confirm #589.

As it is 'shipped' with MXE, I think it is worth to have a closer look. 
The ASCII issue of Project Nr 1 presented to you:
 --- 8< ---
/share/mxe/tools/copydlldeps.py /share/test/ -C /var/lib/jenkins/jobs/DXClusterTelnet/workspace/BIT/64/OS/windows/code/release/ -L /share/mxe/usr/x86_64-w64-mingw32.shared/qt5/bin/
Traceback (most recent call last):
  File "/share/mxe/tools/copydlldeps.py", line 155, in <module>
    for dependency_dll in get_imports(pe_filename_full):
  File "/share/mxe/tools/copydlldeps.py", line 91, in get_imports
    dll_names.append(name.decode("ascii"))
UnicodeDecodeError: 'ascii' codec can't decode byte 0x89 in position 1: ordinal not in range(128)
 --- 8< ---
The second issue is known, output is similar to #589.

I am running mxe on a debian 64bit (intel chipset). Any hints?

**Comment by TimothyGu** [comment]
@dl5rcw you might find https://gist.github.com/TimothyGu/62d8732d914e5db412a5 to be helpful if you are having troubles with the Python version.

**Comment by dl5rcw** [comment]
@TimothyGu Thanks for pointing that out. I picked it up a couple of days ago. I will introduce my version once I am happy with my results here.

**Comment by dl5rcw** [comment]
It took a while until I noticed that dlls from plugin/platforms were not recognized. I thought those would be found by objdump. They are not. Hence, I decided to add a --enforce option to the script in order to add such directories. It works for my 32 and 64 bit builds really well now and started in my production setup last Wednesday. I am quite happy and hope others will like it, too.

My version is based on copydlldeps.sh from @TimothyGu that he pointed out. Thanks, it was a great way to start out! I would like to contribute my version to MXE and therefore wrote explanations, adding all available features/options/parameters. I also wrote a copydlldeps.txt about how I added it in my build scripts. 
  https://github.com/dl5rcw/mxe/blob/master/tools/copydlldeps.sh
